src/snowflake/snowpark/dataframe.py

```diff
@@ -3364,6 +3365,78 @@ def describe(self, *cols: Union[str, List[str]]) -> "DataFrame":
         )
         return res_df
 
+    @df_api_usage
+    def rename(
+        self,
+        col_or_mapper: Union[ColumnOrName, dict],
+        new_column: str = None,
```

Comment:
You could use `@overload` if you want to support two methods of input

Reply:
I think the `@overload` keyword is useful when the number of params are the same (it's used to mention relationships between types of inputs and return) https://adamj.eu/tech/2021/05/29/python-type-hints-how-to-use-overload/

I can write two separate rename functions instead (that would work because the number of parameters are different).

Reply:
The `df.rename(dict, "new_col")` case is already covered by `with_column_renamed` function; which checks if the first parameter passed is a valid column or column name.

`session.create_dataframe([Row(1, 2, 3)], schema=['a', 'b', 'c']).rename({'a':'new_a'}, 'new_a').collect()`
raises
`TypeError: {'a': 'new_a'} must be a column name or Column object.`